File: TelerikUIExamples/TelerikUIExamples/CustomAnimationPieChart.m
//
//  CustomAnimationPieChart.m
//  TelerikUIExamples
//
//  Copyright (c) 2013 Telerik. All rights reserved.
//

#import "CustomAnimationPieChart.h"
#import <TelerikUI/TelerikUI.h>

@interface CustomAnimationPieChart() <TKChartDelegate>
@end

@implementation CustomAnimationPieChart
{
    TKChart *_chart;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self addOption:@"Animate" selector:@selector(animate)];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _chart = [[TKChart alloc] initWithFrame:self.view.bounds];
    _chart.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _chart.allowAnimations = YES;
    _chart.delegate = self;
    [self.view addSubview:_chart];
    
    NSMutableArray *array = [[NSMutableArray alloc] init];
    [array addObject:[[TKChartDataPoint alloc] initWithName:@"Apple" value:@10]];
    [array addObject:[[TKChartDataPoint alloc] initWithName:@"Google" value:@20]];
    [array addObject:[[TKChartDataPoint alloc] initWithName:@"Microsoft" value:@30]];
    [array addObject:[[TKChartDataPoint alloc] initWithName:@"IBM" value:@5]];
    [array addObject:[[TKChartDataPoint alloc] initWithName:@"Oracle" value:@8]];
    
    TKChartPieSeries *series = [[TKChartPieSeries alloc] initWithItems:array];
    series.selection = TKChartSeriesSelectionDataPoint;
    [_chart addSeries:series];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)animate
{
    [_chart animate];
}

#pragma mark TKChartDelegate

// >> chart-anim-pie
- (CAAnimation *)chart:(TKChart *)chart animationForSeries:(TKChartSeries *)series withState:(TKChartSeriesRenderState *)state inRect:(CGRect)rect
{
    CFTimeInterval duration = 0;
    NSMutableArray *animations = [[NSMutableArray alloc] init];
    for (int i = 0; i<state.points.count; i++) {
        NSString *pointKeyPath = [state animationKeyPathForPointAtIndex:i];
        
        NSString *keyPath = [NSString stringWithFormat:@"%@.distanceFromCenter", pointKeyPath];
        CAKeyframeAnimation *a = [CAKeyframeAnimation animationWithKeyPath:keyPath];
        a.values = @[ @50, @50, @0 ];
        a.keyTimes = @[ @0, @(i/(i+1.)), @1 ];
        a.duration = 0.3 * (i+1.1);
        [animations addObject:a];
        
        keyPath = [NSString stringWithFormat:@"%@.opacity", pointKeyPath];
        a = [CAKeyframeAnimation animationWithKeyPath:keyPath];
        a.values = @[ @0, @0, @1 ];
        a.keyTimes = @[ @0, @(i/(i+1.)), @1 ];
        a.duration = 0.3 * (i+1.1);
        [animations addObject:a];
        
        duration = a.duration;
    }
    CAAnimationGroup *g = [[CAAnimationGroup alloc] init];
    g.duration = duration;
    g.animations = animations;
    return g;
}
// << chart-anim-pie

@end

File: Examples/Examples/Chart/Animations/CustomAnimationPieChart.cs
using System;
using System.Collections.Generic;

using Foundation;
using UIKit;
using CoreAnimation;
using CoreGraphics;

using TelerikUI;

namespace Examples
{
	[Register("CustomAnimationPieChart")]
	public class CustomAnimationPieChart: XamarinExampleViewController
	{
		TKChart chart;
		ChartDelegate chartDelegate =  new ChartDelegate ();

		public override void ViewDidLoad ()
		{
			this.AddOption ("Animate", animate);

			base.ViewDidLoad ();

			chart = new TKChart (this.View.Bounds);
			chart.AutoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight;
			chart.Delegate = chartDelegate;
			chart.AllowAnimations = true;
			this.View.AddSubview (chart);

			List<TKChartDataPoint> list = new List<TKChartDataPoint> ();
			list.Add (new TKChartDataPoint(new NSNumber(10), null, "Apple"));
			list.Add (new TKChartDataPoint(new NSNumber(20), null, "Google"));
			list.Add (new TKChartDataPoint(new NSNumber(30), null, "Microsoft"));
			list.Add (new TKChartDataPoint(new NSNumber(5), null, "IBM"));
			list.Add (new TKChartDataPoint(new NSNumber(8), null, "Oracle"));

			TKChartPieSeries series = new TKChartPieSeries (list.ToArray());
			series.Selection = TKChartSeriesSelection.DataPoint;
			chart.AddSeries (series);
		}
			
		public void animate()
		{
			chart.Animate ();
		}

		class ChartDelegate: TKChartDelegate
		{
			// >> chart-anim-pie-cs
			public override CAAnimation AnimationForSeries (TKChart chart, TKChartSeries series, TKChartSeriesRenderState state, CGRect rect)
			{
				double duration = 0;
				List<CAAnimation> animations = new List<CAAnimation>();
				for (int i = 0; i<(int)state.Points.Count; i++) {
					string pointKeyPath = state.AnimationKeyPathForPointAtIndex ((uint)i);
	
					string keyPath = string.Format("{0}.distanceFromCenter", pointKeyPath);
					CAKeyFrameAnimation a = CAKeyFrameAnimation.GetFromKeyPath(keyPath);
					a.Values = new NSNumber[] { new NSNumber(50), new NSNumber(50), new NSNumber(0) };
					a.KeyTimes = new NSNumber[] { new NSNumber(0), new NSNumber(i/(i+1.0)), new NSNumber(1) };
					a.Duration = 0.3 * (i+1.1);
					animations.Add(a);
	
					keyPath = string.Format("{0}.opacity", pointKeyPath);
					a = CAKeyFrameAnimation.GetFromKeyPath(keyPath);
					a.Values = new NSNumber[] { new NSNumber(0), new NSNumber(0), new NSNumber(1) };
					a.KeyTimes = new NSNumber[] { new NSNumber(0), new NSNumber(i/(i+1.0)), new NSNumber(1) };
					a.Duration = 0.3 * (i+1.1);
					animations.Add(a);
	
					duration = a.Duration;
				}
				CAAnimationGroup g = new CAAnimationGroup();
				g.Duration = duration;
				g.Animations = animations.ToArray();
				return g;
			}
			// << chart-anim-pie-cs
		}
	}
}

